Basket-Donation-Payments/Basket-Donation-Payments/View Controllers/AddBasketViewController.m

```diff
@@ -31,15 +35,51 @@ - (IBAction)nextButtonTapped:(id)sender {
     //switch tab controller to index 0
 }
 
-/*
-#pragma mark - Navigation
+- (IBAction)headerImageButtonTapped:(id)sender {
+        UIImagePickerController *imagePickerVC = [UIImagePickerController new];
+        imagePickerVC.delegate = self;
+        imagePickerVC.allowsEditing = YES;
+        BOOL cameraSourceAvail = [UIImagePickerController isSourceTypeAvailable:UIImagePickerControllerSourceTypeCamera];
+        BOOL photoLibAvail = [UIImagePickerController isSourceTypeAvailable:UIImagePickerControllerSourceTypePhotoLibrary];
+        if (cameraSourceAvail && photoLibAvail) {
+            //UIActionSheet to pick a source type
+            UIAlertController *sourceTypePicker = [UIAlertController alertControllerWithTitle:@"" message:@"" preferredStyle:UIAlertControllerStyleActionSheet];
+            UIAlertAction *camera = [UIAlertAction actionWithTitle:@"Camera" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
+                imagePickerVC.sourceType = UIImagePickerControllerSourceTypeCamera;
+            }];
+            UIAlertAction *photoLib = [UIAlertAction actionWithTitle:@"Photo Library" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
+                imagePickerVC.sourceType = UIImagePickerControllerSourceTypePhotoLibrary;
+            }];
+            [sourceTypePicker addAction:camera];
+            [sourceTypePicker addAction:photoLib];
+            [self presentViewController:sourceTypePicker animated:YES completion:nil];
+        } else {
+            imagePickerVC.sourceType = UIImagePickerControllerSourceTypePhotoLibrary;
+        }
+        [self presentViewController:imagePickerVC animated:YES completion:nil];
+    }
+
+- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingMediaWithInfo:(NSDictionary<NSString *,id> *)info {
+    UIImage *originalImage = info[UIImagePickerControllerOriginalImage];
+    UIImage *editedImage = info[UIImagePickerControllerEditedImage];
+
+    self.basketHeaderImageView.image = editedImage;
+    [self dismissViewControllerAnimated:YES completion:nil];
+}
 
-// In a storyboard-based application, you will often want to do a little preparation before navigation
 - (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
-    // Get the new view controller using [segue destinationViewController].
-    // Pass the selected object to the new view controller.
+    if ([segue.identifier isEqualToString:@"addNonprofitsSegue"]) {
+
+        NSDictionary *dict = @{
```

Comment:
You should always use model classes instead of dictionaries, in this case you already have the `Basket` model class which you can use.

Reply:
I think in this case, I was using a dictionary to instantiate a new basket. Is there a better way that I can do this?

Reply:
Yes, you would usually define an `initWith...` method in the `Basket` class that takes all the (required) properties as parameters, that way you enforce type safety which can't be done with an `NSDictionary`. You can also do just `Basket *basket = [Basket new];` and set the properties one by one after that, this way you still have the compiler checking that all the types are correct. An `initWithDictionary` is common when creating a model from a JSON response, which is not type safe, but when you can enforce type safety and use the compiler's help then that's preferred.

Reply:
I see, makes sense! Making that change now  
[](https://github.com/kristiehuang/Basket-Donation-Payments/commit/f55da245a9d30e3607690c2a6f8c139de98bb5b6)